### DISCUSSION TITLE
Bug or feature? Module evaluated multiple times

### DISCUSSION THREAD

**Discussion by elendirx** [vercel/next.js#16301] [Help]
In my API handler (`pages/api/someresource.ts`) I import a module `db.ts` with database connection stored in a module-scoped variable. It is a common thing to do if you need to share a single instance of a class across your app. You can import the module many times, but the connection is only created once when the module is evaluated (=imported for the first time).

I found that with `next dev` this is not true. The `db.ts` module is being evaluated on each request and therefore each time new connection is created.

I wonder if that is expected behavior? Is there another way to cache stuff in memory across API requests?

**Comment by jamesmosier** [comment]
Most of the database connection examples open a connection as it is needed and then close it afterwards. They do not persist the connection. For example the [MongoDB](https://github.com/vercel/next.js/tree/canary/examples/with-mongodb) example works like this.

Although there is another [guide](https://vercel.com/guides/deploying-next-and-mysql-with-vercel) that uses [serverless-mysql](https://github.com/jeremydaly/serverless-mysql) which may work best for your scenario.

**Reply by elendirx** [reply]
Thanks for your message. I spent whole day examining this so I already know all the links. Unfortunately closing the connection at the end of each request brings more headaches than it solves (error may prevent code reaching the `connection.close()` statement, your whole app is limited by database `max_connections` and in general **it's a bad practice**).

True, I can use something like `serverless-mysql`, but why if all I desire is a single **module variable working as expected**?

There are **use cases other than db connection**. How about in-memory caching? Often you simply need to share data across requests and the module-scope variable is very common approach to achieve that.

This is crazy, I don't believe the framework is so poorly implemented, there must be some problem between my chair and my keyboard.

**Comment by elendirx** [comment]
I discovered that during `next dev` two thing happen to the API page:
- it's recompiled every `maxInactiveAge` miliseconds - this is OK as it's documented and available in config
- it's re-imported/re-evaluated every time **some other page** (e.g. `http://localhost:3000/foo`) gets compiled - this is a problem and I don't understand why it happens, maybe it's HMR related.

This is only true for the dev server, in production there's no problem.

For now I ended up checking if the connection already exists (which sucks).

**Reply by crussell42** [reply]
I see the same issue in dev. Hate to say it but this breaks javascript its self. If i cant maintain state across api calls as per  node spec the dev mode is dead to me. Rather just write a standard node express api.

**Comment by pragmasoft-ua** [comment]
I think it is so frequently discussed problem and still not solved, so it worth to be mentioned in documentation
